content/blog/scheduling-in-react/index.md

```diff
@@ -0,0 +1,243 @@
+
+All these parallel steps can lead to janky interfaces and unhappy users, so let’s learn how we can fix this.
+
+## Scheduling in User Interfaces
+
+Our users expect immediate feedback. Whether they are clicking on a button to open a modal or adding text to an input field, they expect some kind of immediate confirmation. The button should show a modal and the input field should display the key that was typed.
+
+To visualize what happens when this is not the case, let’s take a look at the demo application that Dan Abramov presented at his talk, [Beyond React 16](https://reactjs.org/blog/2018/03/01/sneak-peek-beyond-react-16.html), at JSConf Iceland 2018.
+
+The application works like this: The more you type into the input below, the more detailed the charts below will get. Since both of the updates (the input element and the chart) run at the same time, the browser has to do so much computation that it will drop some frames. This leads to noticeable delays and a bad user experience:
+
+<video src="/blog/scheduling-in-react/sync-mode.mp4" muted="true" autoplay muted playsinline loop></video>
+
+However, in this example, it’s more important to update the input with the new text than it is to update the charts. A version that prioritizes the input will appear a lot more responsive to the end user even though the same computation time is required:
+
+<video src="/blog/scheduling-in-react/concurrent-mode.mp4" muted="true" autoplay muted playsinline loop></video>
+
+Unfortunately, current user interface architectures makes it non-trivial to implement this kind of prioritization. One way to solve this is by [debouncing](https://davidwalsh.name/javascript-debounce-function) the chart update. The problem with this is that the charts still render synchronously when the debounced callback fires, which will again cause our user interface to take multiple seconds during which it is not responsive. We can do better!
+
+## Browser Event Loop
+
+Before we learn more about how this can be achieved, let’s dig deeper and understand why the browser has issues with these kind of user interactions.
+
+JavaScript code is executed in one thread, meaning that only one line of JavaScript can be run at any given time. The same thread is also responsible for other document lifecycles like layout and paint.[^1] This means that whenever JavaScript code runs, the browser is blocked from doing anything else.
+
+To keep the user interface responsive, we only have a very short timeframe before we need to be able to receive the next input events. In the browser run loop visualization presented in Subhie Panicker’s and Jason Miller’s talk, [A Quest to Guarantee Responsiveness](https://developer.chrome.com/devsummit/schedule/scheduling-on-off-main-thread), at the Chrome Dev Summit 2018 below, we can see that we only have 16 milliseconds (on a typical 60Hz screen) before the next frame is drawn and the next events need to be processed:
```

Comment:
Also I'd probably break this down into two sentences, as the construction is really awkward and the whole thing is maybe too packed.

"In the browser run loop visualization presented in Subhie Panicker’s and Jason Miller’s talk, [A Quest to Guarantee Responsiveness](https://developer.chrome.com/devsummit/schedule/scheduling-on-off-main-thread), at the Chrome Dev Summit 2018 below, we can see that we only have 16 milliseconds (on a typical 60Hz screen) before the next frame is drawn and the next events need to be processed:"

Something like:

At the Chrome Dev Summit 2018, Subhie Panicker and Jason Miller gave a talk, [A Quest to Guarantee Responsiveness](https://developer.chrome.com/devsummit/schedule/scheduling-on-off-main-thread). During the talk, they presented.......

WDYT?

Reply:
I also think the attribution here is awkward but I felt really bad putting it as a footnote so I left it inline. Splitting it in two sentences sounds good. Your approach seems fine.

Reply:
OK here's the rewrite incorporate the question before:

To keep the user interface responsive, we only have a very short timeframe before we need to be able to receive the next input events. At the Chrome Dev Summit 2018, Shubhie Panicker and Jason Miller gave a talk, [A Quest to Guarantee Responsiveness](https://developer.chrome.com/devsummit/schedule/scheduling-on-off-main-thread). During the talk, they presented the browser's run loop visualization, in which we can see that we only have 16ms (on a typical 60Hz screen) before the next frame is drawn and the next events need to be processed:

Sound good?